mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxFactory.kt

```diff
@@ -1,10 +1,113 @@
 package com.airbnb.mvrx
 
+import android.os.Bundle
+import android.os.Parcelable
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.ViewModelProvider
 import androidx.annotation.RestrictTo
+import java.io.Serializable
+import java.lang.IllegalStateException
+import kotlin.reflect.full.primaryConstructor
 
 @RestrictTo(RestrictTo.Scope.LIBRARY)
-class MvRxFactory<V>(private val factory: (Class<*>) -> V) : ViewModelProvider.Factory {
-    @Suppress("UNCHECKED_CAST") override fun <T : ViewModel?> create(modelClass: Class<T>) = factory(modelClass) as T
-}
\ No newline at end of file
+
+class MvRxFactory<VM : BaseMvRxViewModel<S>, S : MvRxState>(
+    private val viewModelClass: Class<out VM>,
+    private val stateClass: Class<out S>,
+    private val viewModelContext: ViewModelContext,
+    private val key: String,
+    private val forExistingViewModel: Boolean,
+    private val initialStateFactory: MvRxStateFactory<VM, S>
+) : ViewModelProvider.Factory {
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+        val (viewModelContext, viewModel) = viewModelContext.savedStateRegistry.consumeRestoredStateForKey(key)?.let { savedStateBundle ->
+            val restoredArgs = savedStateBundle.get(KEY_MVRX_SAVED_ARGS)
```

Comment:
might be nice to pull the code in this `let` block into a separate function, since this create function is a bit large and complex.